### DISCUSSION TITLE
on premise setup, fhir profiles and bots

### DISCUSSION THREAD

**Discussion by JulianStitz** [medplum/medplum#2483] [General]
hello,

i'd like to use medplum as the base for an on-premise healthcare solution in a german niche market.
therefore i'm interested in using bots in an on-premise way too. is this possible?

we need to integrate a bunch of german profiles. what are the necessary steps to do so.
for example: i'd like to use the management app to create a german "kbv patient": https://simplifier.net/packages/de.dit-connectathon.r4/0.3.0/files/247877/~json

by now you are supporting r4. is there a specific reason why you haven't adopted r5 yet?

ps.: please forgive me if these questions come across as critical. they are not meant to be negative. i would like to engage in a discussion with you to understand what is possible and what is not, and to grasp your motivations behind certain decisions.

**Comment by reshmakh** [comment]
welcome @JulianStitz!  Questions are great.

As for profiles, their official support is a [work in progress](https://github.com/medplum/medplum/discussions/2006) - though current implementations use our tools (SDK, bots, subscriptions, etc) at present to implement profiles.  We have several building blocks, e.g. validation in place for fully integrated profile support coming.

For bots on-premise, not officially supported, major consideration is the sandboxing that comes with lambdas.  Can you share more details about the on-premise environment you'll be in?

Regarding R5 - major downstream systems (payors, other EHRs) do not yet have support yet, and so we haven't seen a lot of activity in implementations on this front as yet.  We have a [post here](https://www.medplum.com/blog/fhir-r5) for background.

Also, if you are interested to join our Discord - we would welcome you there as well.  https://discord.gg/medplum

**Reply by JulianStitz** [reply]
Thank you for your prompt response. I have directly joined your Discord. FHIR profiles are very important to us. As I mentioned, we need to support various profiles. Some are externally defined, while others we have to create ourselves because our domain has been only inadequately, if at all, specified in FHIR.

My current understanding of Medplum is that during the build process, Medplum generates TypeScript types from handcrafted profile JSON files, each containing several StructureDefinitions. These types then serve as a foundation for other components like "App" or "Server" and are also utilized to build widgets. However, this seems to imply that currently, we would need to create a fork to modify these JSON files to suit our specific purposes. It feels like there might be some missing pieces in my understanding.

The operational mode at our customers requires everything to work locally. This means that bots would have to run in local containers within the practices' intranet. How many changes would we need to contribute to enable Medplum for this purpose? Would such an extension align with Medplum's vision?

**Reply by codyebberson** [reply]
Hi @JulianStitz 

### 1. On-prem bots

We recently re-released on-prem bots: https://github.com/medplum/medplum/pull/2528

We're working on a documentation push + fit and finish to polish it up, but that will enable running bots on the same host as the server.

### 2. FHIR Profiles

> My current understanding of Medplum is that during the build process, Medplum generates TypeScript types from handcrafted profile JSON files, each containing several StructureDefinitions. These types then serve as a foundation for other components like "App" or "Server" and are also utilized to build widgets. However, this seems to imply that currently, we would need to create a fork to modify these JSON files to suit our specific purposes. It feels like there might be some missing pieces in my understanding.

That only applies to the base `StructureDefinition` resources.  Our intent with FHIR Profiles is for it to be additive.

For example, say you had a `Patient` resource that you wanted to use with your custom profile.

To do that, you will set the profile:

```js
{
  "resourceType": "Patient",
  "meta": {
    "profile": [
      "https://fhir.kbv.de/StructureDefinition/KBV_PR_Base_Patient"
    ]
  },
  // ...
}
```

And then the server will enforce all of the profile constraints at runtime.

For your developer experience, you could continue to use the `Patient` type definition from `@medplum/fhirtypes`.  Or you could use your own type definitions, which are more strict but still compatible with `Patient`.

We have tools that generate TypeScript definitions from StructureDefinitions, but that's only an internal tool for now.  We'd be happy to discuss prioritizing the work to polish that tool, and to properly release it.
